### PR TITLE
Add Find-my-Friends Service

### LINKED ISSUE #123: Find My Friends Support?
Hi,

Is there plans to add find my friends support to PyiCloud?

Thanks

### LINKED ISSUE #231: PyiCloud Find-My-Friends Integration (solved)
I created a customized version of PyiCloud that adds the Find-my-Friends service interface between PyiCloud and Apple's iCloud Web Services using code provided in several older issues and other web sites. This customized version has been in use by the Home Assistant iCloud3 custom component since September, 2019. (iCloud3 provides enhanced device tracking for Home Assistant users using the Find-my-Friends service to create a link between a non-2fa/non-2sa account and a 2fa account.)

Now that PyiCloud is back on an active maintenance path, I wanted to bring this code to your attention in case you want to add this service to PyiCloud. Upon review, you will find several new Classes that might be able to be easily incorporated into a new Find-my-Friends external service similar to the others provided by PyiCloud. Attached is the pyicloud_ic3.py code used by iCloud3. 

FYI - Links to the iCloud3 GitHub repository:
* iCloud3 GitHub Repository is [here](https://github.com/gcobb321/icloud3)
* iCloud3 documentation is [here](https://gcobb321.github.io/icloud3/#/)

Gary Cobb

[pyicloud_ic3.zip](https://github.com/picklepete/pyicloud/files/4238212/pyicloud_ic3.zip)

### PR DESCRIPTION
**Pull request by gcobb321**
## Proposed change
Add the Find-my-Friends service

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New service (thank you!)
- [x] New feature (which adds functionality to an existing service)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Documentation or code sample

## Example of code:
See README.rst

## Additional information
Adds support to locate devices, access contact information and access a list of followed contacts for the account.

- This PR fixes or closes issue: fixes #123 & fixes #231 & closes #160
- This PR is related to issue: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated to README

### CODE REVIEW

**Comment by gcobb321** [comment]
The findmyfriends.py module has been tested with the iCloud3 Device Tracker.

File: pyicloud/base.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 )
 from pyicloud.services import (
     FindMyiPhoneServiceManager,
+    FindMyFriendsServiceManager,
```
Comment (Quentame) [inline pyicloud/base.py:21]:
```suggestion
    FindMyFriendsService,
```

FindMyiPhone service will be renamed in ##227 :wink:

File: pyicloud/base.py [hunk 2]
```diff
@@ -334,6 +335,16 @@ def account(self):
         service_root = self._get_webservice_url("account")
         return AccountService(service_root, self.session, self.params)
 
+    @property
+    def friends(self):
+        """Gets the 'Find-my-Friends' service."""
+        service_root = self._get_webservice_url('fmf')
+        return FindMyFriendsServiceManager(
```
Comment (Quentame) [inline pyicloud/base.py:342]:
```suggestion
        return FindMyFriendsService(
```

File: pyicloud/services/findmyfriends.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return self.data.get('contactDetails')
+
+    @property
+    def followers(self):
+        return self.data.get('followers')
+
+    @property
+    def following(self):
+        return self.data.get('following')
+
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:72]:
```suggestion
    @property
    def friend_fences(self):
        return self.data.get('friendFencesISet')

    @property
    def my_fences(self):
        return self.data.get('myFencesISet')

```

File: pyicloud/services/findmyfriends.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                'windowVisible': True
+            },
+            'dataContext': None,
+            'serverContext': None
+        })
+        req = self.session.post(
+            self._friend_endpoint,
+            data=requested_data,
+            params=params
+            )
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:46]:
```suggestion
        )
```

File: README.rst [hunk 5]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
 >>> api.iphone.lost_device(phone_number, message)
 
 
+Find-my-Friends
+===============
+The Find-my-Friends service is lets you locate the friends you have you are following using the FindMe app. Once you have successfully authenticated, you can request the friends you are following and their location.
```
Comment (Quentame) [inline README.rst:163]:
```suggestion
The Find-my-Friends service lets you locate the friends you are following using the FindMe app.
```

Too complicated + repetitions

File: README.rst [hunk 6]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
...
+
+
+Friends
+*******
+Returns the Find-my-Friends object.
```
Comment (Quentame) [inline README.rst:168]:
```suggestion
Returns the Find-my-Friends service.
```

File: README.rst [hunk 7]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
...
+
+>>> fmf = api.friends
+
+The fmf friends contact, location and following data is structured with the following dictionaries:
```
Comment (Quentame) [inline README.rst:172]:
```suggestion
The `contacts`, `following` and `locations` data is structured with the following dictionaries:
```

File: README.rst [hunk 8]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
...
+Returns a dictionary of all of the contacts that are being followed. They have accepted a following invitation request from the account.
+
+
+>>> following = fmf.following
+
+Return the ID number and email address of the contact being followed. 
+
+>>> contact_id = following['id']
+>>> contact_email = following['invitationAcceptedHandles']
+
```
Comment (Quentame) [inline README.rst:193]: Can you add the`fullName` ?

File: README.rst [hunk 9]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
...
+
+Contacts
+********
+Returns a dictionary of all contacts for the account.
+
+>>> contacts = fmf.contacts
+
+
+Following Contacts
```
Comment (Quentame) [inline README.rst:182]:
```suggestion
Following contacts
```
No upper case in the middle of a sentence/title :relieved:

File: README.rst [hunk 10]
```diff
@@ -158,6 +158,49 @@ Lost mode is slightly different to the "Play Sound" functionality in that it all
...
+
+>>> following = fmf.following
+
+Return the ID number and email address of the contact being followed. 
+
+>>> contact_id = following['id']
+>>> contact_email = following['invitationAcceptedHandles']
+
+
+Location of Followed Friends
```
Comment (Quentame) [inline README.rst:195]:
```suggestion
Location of followed friends
```

File: pyicloud/services/findmyfriends.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import json
+import sys
+import six
+
+from pyicloud.exceptions import PyiCloudNoDevicesException
+
+class FindMyFriendsServiceManager(object):
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:7]:
```suggestion
class FindMyFriendsService(object):
```

File: pyicloud/services/findmyfriends.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
+import json
+import sys
+import six
+
+from pyicloud.exceptions import PyiCloudNoDevicesException
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:5]:
```suggestion
```

Not used

File: pyicloud/services/findmyfriends.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
+import json
+import sys
+import six
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:3]:
```suggestion
```

Not used

File: pyicloud/services/findmyfriends.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
+import json
+import sys
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:2]:
```suggestion
```

Not used

File: pyicloud/services/findmyfriends.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
+import json
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:1]: Missing module docstring

File: pyicloud/services/findmyfriends.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+            'serverContext': None
+        })
+        req = self.session.post(
+            self._friend_endpoint,
+            data=requested_data,
+            params=params
+            )
+        self.response = req.json()
+        
+        return self.response
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:49]: You are not using response outside of this method.

```suggestion
        return req.json()
```

File: pyicloud/services/findmyfriends.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+            },
+            'dataContext': None,
+            'serverContext': None
+        })
+        req = self.session.post(
+            self._friend_endpoint,
+            data=requested_data,
+            params=params
+            )
+        self.response = req.json()
```
Comment (Quentame) [inline pyicloud/services/findmyfriends.py:47]:
```suggestion
```

File: pyicloud/services/findmyfriends.py (hunk 16 above)
Comment (zzeleznick) [reply pyicloud/services/findmyfriends.py:49]: @Quentame - It looks like the author here has intentionally copied my code almost verbatim without attribution (e.g. forking my code like a good citizen).

I would give the author the benefit of doubt he forked off my fork (or mentioned https://github.com/picklepete/pyicloud/pull/160/fileshttps://github.com/picklepete/pyicloud/pull/160/files), but the author here even used my original method docstrings.

**Comment by elzii** [comment]
Note: This is not as easy in macOS 15+, since the items in `~/Library/Caches/com.apple.findmy.fmipcore/` are now encoded. 
You can get them more readable in XML with `sudo plutil -convert xml1 -o ~/Downloads/Items.txt ./Items.data` but it's still an unknown.

The hunt is on to find how to decrypt them.  See comments on this [gist](https://gist.github.com/henrik242/1da3a252ca66fb7d17bca5509a67937f?permalink_comment_id=5321162#gistcomment-5321162)

And ask Gemini/Ollama/GPT-4 etc about using Darwin's `security` and `plutil`.  I was able to find my apple user ID and password, which are then Base64 encoded for Basic Auth, in `com.apple.account.AppleAccount.cloudkit-token`. 

Also, try `binwalk` a copy of the `/usr/libexec/fmfd` (find my friends daemon) with `binwalk --dd='.*' fmfd` to extract all.

Anyone good at reading hex and byte offsets can make work which certs it's using. Example output without the --dd:

```
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Mach-O universal binary with 2 architectures
20364         0x4F8C          Unix path: /usr/lib/dyld
24456         0x5F88          Unix path: /usr/lib/swift/libswiftObjectiveC.dylib
647500        0x9E14C         XML document, version: "1.0"
778255        0xBE00F         mcrypt 2.2 encrypted data, algorithm: blowfish-448, mode: CBC, keymode: 8bit
857010        0xD13B2         XML document, version: "1.0"
863590        0xD2D66         Certificate in DER format (x509 v3), header length: 4, sequence length: 1038
864632        0xD3178         Certificate in DER format (x509 v3), header length: 4, sequence length: 1204
865840        0xD3630         Certificate in DER format (x509 v3), header length: 4, sequence length: 1211
866515        0xD38D3         Certificate in DER format (x509 v3), header length: 4, sequence length: 260
867417        0xD3C59         XML document, version: "1.0"
889036        0xD90CC         Unix path: /usr/lib/dyld
893128        0xDA0C8         Unix path: /usr/lib/swift/libswiftObjectiveC.dylib
1563692       0x17DC2C        XML document, version: "1.0"
1728599       0x1A6057        mcrypt 2.2 encrypted data, algorithm: blowfish-448, mode: CBC, keymode: 8bit
1758178       0x1AD3E2        XML document, version: "1.0"
1764758       0x1AED96        Certificate in DER format (x509 v3), header length: 4, sequence length: 1038
1765800       0x1AF1A8        Certificate in DER format (x509 v3), header length: 4, sequence length: 1204
1767008       0x1AF660        Certificate in DER format (x509 v3), header length: 4, sequence length: 1211
1767683       0x1AF903        Certificate in DER format (x509 v3), header length: 4, sequence length: 260
1768585       0x1AFC89        XML document, version: "1.0"
```

**Pull request closed without merging**